metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -892,48 +909,62 @@ class MigrationWriteOperation implements ControllerWriteOperation<Void> {
             }
             @Override
             public ControllerResult<Void> generateRecordsAndResult() {
-                return ControllerResult.atomicOf(batch, null);
+                return ControllerResult.of(batch, null);
             }
 
             public void processBatchEndOffset(long offset) {
                 highestMigrationRecordOffset = new OffsetAndEpoch(offset, curClaimEpoch);
             }
         }
         @Override
-        public void beginMigration() {
+        public CompletableFuture<?> beginMigration() {
             log.info("Starting ZK Migration");
-            // TODO use KIP-868 transaction
+            ControllerWriteEvent<Void> batchEvent = new ControllerWriteEvent<>(
```

Comment:
We need to check the metadata version, right?

If it's too old we just have to return the empty list, I guess. And probably log a WARN message about how transactions aren't available.